File: 2-add-two-numbers/add-two-numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        # dummy = ListNode(0)
        # tail = dummy
        # carry = 0
        # while l1 is not None or l2 is not None or carry != 0:
        #     d1 = l1.val if l1 is not None else 0
        #     d2 = l2.val if l2 is not None else 0
        #     s = d1 + d2 + carry
        #     carry = s // 10
        #     s = s % 10

        #     sum_node = ListNode(s)
        #     tail.next = sum_node
        #     tail = tail.next

        #     l1 = l1.next if l1 is not None else None
        #     l2 = l2.next if l2 is not None else None
        
        # ans = dummy.next
        # dummy.next = None
        # return ans

        dummy = ListNode()
        ans = dummy
        total = carry = 0
        while l1 or l2 or carry:
            total = carry

            if l1:
                total += l1.val
                l1 = l1.next
            if l2:
                total += l2.val
                l2 = l2.next
            
            carry = total // 10
            total %= 10
            dummy.next = ListNode(total)
            dummy = dummy.next
        return ans.next

File: 2-add-two-numbers/add-two-numbers.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = function(l1, l2) {
    var carry = 0, ans = new ListNode(0), curr = ans;
    while(l1 !== null || l2 !== null || carry !== 0) {
        let x = l1 !== null ? l1.val : 0;
        let y = l2 !== null ? l2.val : 0;
        var summ = x + y + carry;
        carry = Math.floor(summ / 10);
        curr.next = new ListNode(summ % 10);
        curr = curr.next;
        if (l1 !== null) l1 = l1.next;
        if (l2 !== null) l2 = l2.next;
    }
    return ans.next;
};